subprojects/core/src/main/java/org/gradle/execution/workgraph/WorkGraphBuilder.java

```diff
@@ -0,0 +1,233 @@
+        markEntryNodesAsShouldRun(dag, requestedNodes);
+        return new WorkGraph(dag, ImmutableMap.copyOf(taskNodes), ImmutableSet.copyOf(requestedNodes), filteredTasks.build());
+    }
+
+    @Nullable
+    private TaskNode addEdge(Graph graph, Map<Task, TaskNode> taskNodes, TaskNode sourceNode, EdgeType type, Task targetTask) {
+        if (filter.isSatisfiedBy(targetTask)) {
+            TaskNode targetNode = getOrCreateTaskNode(graph, taskNodes, targetTask);
+            if (targetNode != null) {
+                graph.addEdge(new Edge(sourceNode, type, targetNode));
+                return targetNode;
+            }
+        }
+        return null;
+    }
+
+    @Nullable
+    private TaskNode addEdge(Graph graph, Map<Task, TaskNode> taskNodes, Task sourceTask, EdgeType type, TaskNode targetNode) {
+        if (filter.isSatisfiedBy(sourceTask)) {
+            TaskNode sourceNode = getOrCreateTaskNode(graph, taskNodes, sourceTask);
+            if (sourceNode != null) {
+                graph.addEdge(new Edge(sourceNode, type, targetNode));
+                return sourceNode;
+            }
+        }
+        return null;
+    }
+
+    @Nullable
+    private TaskNode getOrCreateTaskNode(Graph graph, Map<Task, TaskNode> taskNodes, Task task) {
+        TaskNode node = taskNodes.get(task);
+        if (node == null) {
+            node = createTaskNodeUnlessFiltered(graph, taskNodes, task);
+        }
+        return node;
+    }
+
+    @Nullable
+    private TaskNode createTaskNodeUnlessFiltered(Graph graph, Map<Task, TaskNode> taskNodes, Task task) {
+        if (!filter.isSatisfiedBy(task)) {
+            return null;
+        }
+        TaskNode node = new TaskNode((TaskInternal) task);
+        graph.addNode(node);
+        taskNodes.put(task, node);
+        return node;
+    }
+
+    private static void connectFinalizerDependencies(final Graph graph) {
+        for (Edge edge : graph.getAllEdges()) {
+            if (edge.getType() != FINALIZED_BY) {
+                continue;
+            }
+            final Set<Edge> edgesAddedFromFinalized = Sets.newHashSet();
+            final Node finalized = edge.getSource();
+            Node finalizer = edge.getTarget();
+            graph.walkIncomingEdgesFrom(finalizer, new Graph.EdgeWalkerAction() {
+                @Override
+                public boolean execute(Edge edge) {
+                    if (edge.getType() != DEPENDENCY_OF) {
+                        return false;
+                    }
+                    Node finalizerDependency = edge.getSource();
+                    if (finalizerDependency == finalized) {
+                        return false;
+                    }
+                    Edge finalizerDependencyConstraint = new Edge(finalized, AVOID_STARTING_BEFORE_FINALIZED, finalizerDependency);
+                    if (edgesAddedFromFinalized.add(finalizerDependencyConstraint)) {
+                        graph.addEdge(finalizerDependencyConstraint);
```

Comment:
Should we return `true` here and return `false` at the end? I don't think we need to look at the source of an edge when we know we already visited that source node - and that is the case here since we added the finalizer dependency edge.

Reply:
Good point.